### PR TITLE
Character encoder and data loading.

### PR DESCRIPTION
**Pull request by SuperLinguini**

### CODE REVIEW

File: train.py [hunk 1]
```diff
@@ -366,15 +407,15 @@ def train():
...
+            (args.batch_size * len(train_data[0]) * args.epochs) / (time.time() - start_train_time)))
 
 if __name__ == '__main__':
     start_pipeline_time = time.time()
     if not args.eval_only:
         train()
     model.collect_params().load(args.save, context)
-    val_L = evaluate(model, val_data, context[0])
-    test_L = evaluate(model, test_data, context[0])
+    val_L = evaluate(model, val_data, context[0]) / 2
```
Comment (cgraywang) [inline train.py]: why divided by 2?

File: train.py [hunk 2]
```diff
@@ -366,15 +407,15 @@ def train():
...
+    test_L = evaluate(model, test_data, context[0]) / 2
```
Comment (cgraywang) [inline train.py]: same as above

File: train.py [hunk 3]
```diff
@@ -72,7 +73,26 @@
...
+    "embedding": {
+      "dim": 16
+    },
+    "n_characters": 262,
+    "max_characters_per_token": 50
+  }
+}
 
 
 class ElmoLSTM(gluon.Block):
```
Comment (cgraywang) [inline train.py]: Make this in a separate model file

File: train.py [hunk 4]
```diff
@@ -115,18 +138,18 @@ def forward(self, inputs, states_forward=None, states_backward=None):
 
         for j in range(self.num_layers):
             outputs_forward.append([])
-            for i in range(inputs.shape[0]):
+            for i in range(seq_len):
```
Comment (cgraywang) [inline train.py]: use inputs.shape[0]

File: train.py [hunk 5]
```diff
@@ -138,8 +161,7 @@ def forward(self, inputs, states_forward=None, states_backward=None):
 
         return outputs_forward, out_states_forward, outputs_backward, out_states_backward
 
-
-class StandardRNN(Block):
+class StandardRNN(gluon.Block):
```
Comment (cgraywang) [inline train.py]: make this in the same model file as the ElmoLSTM

File: train.py [hunk 6]
```diff
@@ -307,56 +338,66 @@ def train():
...
-                    L = L + batch_L.as_in_context(context[0]) / X.size
-                    Ls.append(batch_L)
+                for j, (X_forward, X_backward, y_forward, y_backward, h) in enumerate(zip(forward_data_list, backward_data_list, forward_target_list, backward_target_list, hiddens)):
+                    output, h = model(X_forward, X_backward, *h)
+                    batch_L = loss(mx.nd.reshape(output[0], (-3, -1)), mx.nd.reshape(y_forward, (-1,)))
+                    L = L + batch_L.as_in_context(context[0]) / y_forward.size
+                    Ls.append(batch_L / y_forward.size)
+
+                    batch_L = loss(mx.nd.reshape(output[1], (-3, -1)), mx.nd.reshape(y_backward, (-1,)))
+                    L = L + batch_L.as_in_context(context[0]) / y_forward.size
```
Comment (cgraywang) [inline train.py]: nice

File: train.py [hunk 7]
```diff
@@ -307,56 +338,66 @@ def train():
...
         hiddens = [model.begin_state(batch_size=args.batch_size//len(context), func=mx.nd.zeros, ctx=ctx) for ctx in context]
-        for i, (data, target) in enumerate(train_data):
-            data_list = gluon.utils.split_and_load(data, context, batch_axis=1, even_split=True)
-            target_list = gluon.utils.split_and_load(target, context, batch_axis=1, even_split=True)
+
+        batch_i, i = 0, 0
+        while i < len(train_data[0]) - 1 - 1:
+            data, target = get_batch(train_data, i, seq_len=args.bptt)
+
+            forward_data_list = gluon.utils.split_and_load(data[0], context, batch_axis=1, even_split=True)
```
Comment (cgraywang) [inline train.py]: why not use the same data list, the forward and backward layer should handle the data order

File: train.py [hunk 8]
```diff
@@ -207,43 +223,55 @@ def _get_decoder(self):
...
+max_word_length = 50
+
+vocab = UnicodeCharsVocabulary(nlp.data.Counter(train_dataset[0]), max_word_length)
 
-train_data, val_data, test_data = [x.bptt_batchify(vocab, args.bptt, args.batch_size,
-                                                   last_batch='keep')
+train_data, val_data, test_data = [x.batchify(vocab, args.batch_size, max_word_length, load='train_data' if x is train_dataset else None)
                                    for x in [train_dataset, val_dataset, test_dataset]]
 
+def get_batch(data_source, i, seq_len=None):
```
Comment (cgraywang) [inline train.py]: use the simple get_batch, let the forward and backward layer handle the order

File: README.md [hunk 9]
```diff
@@ -1 +1,3 @@
-# ELMo
\ No newline at end of file
+# ELMo
+
+To run: `python train.py [--gpus 0,1,2,3]`
```
Comment (cgraywang) [inline README.md]: what is the current result?

File: data.py [hunk 10]
```diff
@@ -0,0 +1,297 @@
...
+    #             has_short_batch = _slice_pad_length(sample_len*batch_size, seq_len+1, 1) > 0
+    #             if has_short_batch:
+    #                 batches.append(data[seq_len*len(batches):, :])
+    #     # TODO: +1 for masks
+    #     return SimpleDataset(batches).transform(lambda x:
+    #                                             (vocab.array_to_char_ids(x[:min(len(x)-1, seq_len), :], max_word_length),
+    #                                              mx.nd.array(vocab[x[1:, :]])))
+
+
+def _slice_pad_length(num_items, length, overlap=0):
```
Comment (cgraywang) [inline data.py]: delete this method, use the one in gluonnlp.data.utils

File: train.py [hunk 11]
```diff
@@ -115,18 +138,18 @@ def forward(self, inputs, states_forward=None, states_backward=None):
...
+                    output, states_forward[j] = self.forward_layers[j](inputs_forward[i], states_forward[j])
                 else:
                     output, states_forward[j] = self.forward_layers[j](outputs_forward[j-1][i], states_forward[j])
                 outputs_forward[j].append(output)
             out_states_forward.append(states_forward[j])
 
-            outputs_backward.append([None] * inputs.shape[0])
-            for i in reversed(range(inputs.shape[0])):
+            outputs_backward.append([None] * seq_len)
+            for i in reversed(range(seq_len)):
```
Comment (cgraywang) [inline train.py]: same here

File: train.py (hunk 5 above)
Comment (cgraywang) [reply train.py]: all we need it ElmoLM, delete this StandardRNN. It has specific implementation in gluonnlp.

File: elmo_char_encoder.py [hunk 12]
```diff
@@ -0,0 +1,164 @@
...
+        return current_input
+
+class ElmoCharacterEncoder(gluon.Block):
+    def __init__(self, options):
+        super(ElmoCharacterEncoder, self).__init__()
+
+        if isinstance(options, dict):
+            self._options = options
+        else:
+            with open(options, 'r') as fin:
```
Comment (cgraywang) [inline elmo_char_encoder.py]: separate the weights loading in a separate method

File: README.md [hunk 13]
```diff
@@ -1 +1,3 @@
-# ELMo
\ No newline at end of file
+# ELMo
```
Comment (cgraywang) [inline README.md]: can you also help to make this repo private? I found this is very similar to what we have implemented: https://github.com/allenai/allennlp/pull/1154

File: train.py (hunk 1 above)
Comment (SuperLinguini) [reply train.py]: It just makes it easier to compare the perplexity to what is expected for WikiText2 by default. The forward and backward LSTMs have nearly identical perplexities.

File: train.py (hunk 8 above)
Comment (SuperLinguini) [reply train.py]: The target for simple get_batch only works for the forward layer. The backward layer requires a different target.

File: LSTMPCell.py [hunk 14]
```diff
@@ -0,0 +1,230 @@
+from mxnet.gluon.rnn import HybridRecurrentCell
+from mxnet import ndarray, nd
+from mxnet.gluon.rnn import LSTM, BidirectionalCell
+
+class LSTMPCell(HybridRecurrentCell):
```
Comment (szha) [inline LSTMPCell.py]: Looks like a copy of https://mxnet.incubator.apache.org/versions/master/api/python/gluon/contrib.html?highlight=lstmpcell#mxnet.gluon.contrib.rnn.LSTMPCell

**Comment by cgraywang** [comment]
This is officially the final week, let's try to catch up with everything left by refering: https://quip-amazon.com/KTy9AubUodb4

File: LSTMPCell.py (hunk 14 above)
Comment (cgraywang) [reply LSTMPCell.py]: We are going to rewrite, thanks for the pointer

File: train.py [hunk 15]
```diff
@@ -167,52 +195,12 @@ def train():
...
-                cur_L = total_L / (args.log_interval * 2)
-                print('[Epoch %d Batch %d/%d] loss %.2f, ppl %.2f, throughput %.2f samples/s, lr %.2f' % (
-                    epoch, batch_i, len(train_data[0]) // args.bptt, cur_L, math.exp(cur_L),
-                    args.batch_size * args.log_interval / (time.time() - start_log_interval_time), args.lr))
-                total_L = 0.0
-                start_log_interval_time = time.time()
-            i += args.bptt
-            batch_i += 1
+
+        if args.char_embedding:
```
Comment (cgraywang) [inline train.py]: now we are using word language model script to train to ensure the alignment between the implentation and preventing from possible bugs in this train script

File: train.py [hunk 16]
```diff
@@ -414,8 +242,8 @@ def train():
     if not args.eval_only:
         train()
     model.collect_params().load(args.save, context)
-    val_L = evaluate(model, val_data, context[0]) / 2
-    test_L = evaluate(model, test_data, context[0]) / 2
+    val_L = evaluate(model, val_data, context[0])
```
Comment (cgraywang) [inline train.py]: the total loss is handled in the evaluation script

File: word_language_model.py [hunk 17]
```diff
@@ -49,6 +49,8 @@
 import gluonnlp as nlp
```
Comment (cgraywang) [inline word_language_model.py]: migrate training into the this script

File: word_language_model.py [hunk 18]
```diff
@@ -110,6 +112,235 @@
...
+
+        lstm_input_size = input_size
+
+        with self.name_scope():
+            for layer_index in range(num_layers):
+                # forward_layer = LSTMPCell(hidden_size, cell_size, input_size=lstm_input_size, memory_cell_clip_value=3, state_projection_clip_value=3)
+                # backward_layer = LSTMPCell(hidden_size, cell_size, input_size=lstm_input_size, memory_cell_clip_value=3, state_projection_clip_value=3)
+
+                # forward_layer = _get_rnn_cell(mode, 1, lstm_input_size, hidden_size, 0 if layer_index == num_layers - 1 else dropout)#, cell_size=cell_size)
+                backward_layer = _get_rnn_cell(mode, 1, lstm_input_size, hidden_size, 0 if layer_index == num_layers - 1 else dropout)#, cell_size=cell_size)
```
Comment (cgraywang) [inline word_language_model.py]: remove forward layer to verify, the backword layer only seems generate higher ppl, it seems align with the language model mathmatical intuition

**Comment by cgraywang** [comment]
generally refer to https://quip-amazon.com/KTy9AubUodb4

File: elmo_char_cnn_train.py [hunk 19]
```diff
@@ -25,6 +25,7 @@
 from mxnet.gluon import nn, Block
 
 from ELMo.data import UnicodeCharsVocabulary, WikiText2Character
```
Comment (cgraywang) [inline elmo_char_cnn_train.py]: The char embedding loader is still under testing

File: word_language_model.py [hunk 20]
```diff
@@ -0,0 +1,405 @@
...
+            L = 0
+            with autograd.record():
+                for j, (X, y, h) in enumerate(zip(data_list, target_list, hiddens)):
+                    output, h, encoder_hs, dropped_encoder_hs = forward(X, h)
+                    l = criterion(output, y, encoder_hs, dropped_encoder_hs)
+                    L = L + l.as_in_context(context[0]) / X.size
+                    Ls.append(l/X.size)
+                    hiddens[j] = h
+            L.backward()
+            grads = [p.grad(d.context) for p in parameters for d in data_list]
```
Comment (cgraywang) [inline word_language_model.py]: verify the gradient exploration issue
